streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -282,7 +284,22 @@ public <K, V> void send(final String topic,
                     topicProducedSensor.record(bytesProduced, context.currentSystemTimeMs());
                 }
             } else {
-                recordSendError(topic, exception, serializedRecord, context, processorNodeId);
+                final RecordContext recordContext = context.recordContext();
+                recordSendError(
+                    topic,
+                    exception,
+                    serializedRecord,
+                    recordContext != null ?
+                        new ProcessorRecordContext(
```

Comment:
Given what the callback is executed on the producer sender thread, I think we need to make a deep copy here. Otherwise, the information inside the `RecordContext` object might change under the hood.

Reply:
I am not sure, whether this makes a difference. Once you get the reference with `context.recordContext()` the record context will not change anymore since it is an immutable object.